lib/shrine/plugins/derivates.rb

```diff
@@ -0,0 +1,151 @@
+# frozen_string_literal: true
+
+class Shrine
+  module Plugins
+    module Derivates
+      def self.load_dependencies(uploader, *)
+        uploader.plugin :default_url
+      end
+
+      def self.configure(uploader, opts = {})
+        uploader.opts[:derivates_attribute] = opts.fetch(:data_attribute, :separate) # values are separate or original
+        uploader.opts[:version_fallbacks] = opts.fetch(:fallbacks, uploader.opts.fetch(:version_fallbacks, {}))
+        uploader.opts[:versions_fallback_to_original] = opts.fetch(:fallback_to_original, uploader.opts.fetch(:versions_fallback_to_original, true))
+
+        unless [:separate, :original].include? uploader.opts[:derivates_attribute]
+          raise Error, "`#{uploader.opts[:derivates_attribute] || "nil"}` is an invalid " \
+            "option for `data_attribute` in Derivates Plugin. Valid options are: " \
+            "[:separate, :original]"
+        end
+      end
+
+      module ClassMethods
+        def version_fallbacks
+          opts[:version_fallbacks]
+        end
+      end
+
+      module AttachmentMethods
+        def initialize(name, **options)
+          super
+
+          module_eval <<-RUBY, __FILE__, __LINE__ + 1
+            def #{name}_derivates
+              #{name}_attacher.get_derivates
+            end
+          RUBY
+        end
+      end
+
+      module AttacherMethods
+        def changed?
+          instance_variable_defined?(:@old) || instance_variable_defined?(:@derivates_queue)
+        end
+
+        def finalize
+          super
+          update_derivates
+        end
+
+        def add_derivate(key, io, store = :store, nested_list = nil)
+          @derivates_queue ||= {}
```

Comment:
`@derivates_queue` should also handle removal of attachments. My vision is that I would like to make [revising derivates](https://shrinerb.com/rdoc/files/doc/regenerating_versions_md.html) easy, and that sometimes requires removing existing derivates.